File: dashboard/src/logic/caseValidator.ts
import type { CaseRecord, ParsedRow } from './types';
import { DateTimeParser } from './dateTimeParser';

export interface ValidatorConfig {
  startPadMinutes: number;
  endPadMinutes: number;
  binCapacity: number;
  maxProcedureDurationHours?: number;
  minProcedureDurationMinutes?: number;
}

export class CaseValidator {
  private processedIds = new Set<string>();
  private config: ValidatorConfig;

  constructor(config: ValidatorConfig) {
    this.config = config;
  }

  validate(rowNumber: number, record: CaseRecord | null, unparseableField?: string): ParsedRow {
    if (unparseableField) {
      return this.createInvalid(rowNumber, record, `Unparseable field: ${unparseableField}`);
    }

    if (!record) {
      return this.createInvalid(rowNumber, null, "Missing data record");
    }

    // Duplicate Check
    if (this.processedIds.has(record.id)) {
      return this.createInvalid(rowNumber, record, "Duplicate Case_ID");
    }
    this.processedIds.add(record.id);

    const anesthStart = DateTimeParser.addMinutes(record.csvStart, -this.config.startPadMinutes);
    const anesthEnd = DateTimeParser.addMinutes(record.csvEnd, this.config.endPadMinutes);

    // Occupancy-based logic for packing/duration
    const occStart = DateTimeParser.addMinutes(record.patientIn, -this.config.startPadMinutes);
    const occEnd = DateTimeParser.addMinutes(record.patientOut, this.config.endPadMinutes);
    const duration = DateTimeParser.minutesBetween(occStart, occEnd);

    // Cross-midnight Check
    if (duration < 0) {
      return this.createInvalid(rowNumber, record, `Cross-midnight occupancy (end before start). In: ${record.patientIn}, Out: ${record.patientOut}`);
    }

    // Min Duration Check
    const minMin = this.config.minProcedureDurationMinutes || 0;
    if (duration < minMin) {
      return this.createInvalid(rowNumber, record, `Occupancy under min duration (<${minMin}m)`);
    }

    // Capacity Check
    const outOfBounds = duration > this.config.binCapacity;
    if (outOfBounds) {
      return {
        rowNumber,
        record,
        anesthesiaStart: anesthStart,
        anesthesiaEnd: anesthEnd,
        occupancyStart: occStart,
        occupancyEnd: occEnd,
        durationMinutes: duration,
        valid: false,
        invalidReason: `Occupancy over max duration (>${this.config.binCapacity}m)`,
        outOfBounds: true
      };
    }

    return {
      rowNumber,
      record,
      anesthesiaStart: anesthStart,
      anesthesiaEnd: anesthEnd,
      occupancyStart: occStart,
      occupancyEnd: occEnd,
      durationMinutes: duration,
      valid: true,
      invalidReason: null,
      outOfBounds: false
    };
  }

  private createInvalid(rowNumber: number, record: CaseRecord | null, reason: string): ParsedRow {
    return {
      rowNumber,
      record,
      anesthesiaStart: null,
      anesthesiaEnd: null,
      occupancyStart: null,
      occupancyEnd: null,
      durationMinutes: 0,
      valid: false,
      invalidReason: reason,
      outOfBounds: false
    };
  }

  reset() {
    this.processedIds.clear();
  }
}

File: src/main/kotlin/org/heailth/reader/CaseValidator.kt
package org.heailth.reader

import org.heailth.config.ConfigLoader
import org.heailth.model.CaseRecord
import org.heailth.model.ParsedRow
import org.slf4j.LoggerFactory
import java.time.LocalTime
import java.time.temporal.ChronoUnit

class CaseValidator(private val config: ConfigLoader) {
    private val seenCaseIds = mutableMapOf<String, CaseRecord>()

    fun validate(rowNum: Long, record: CaseRecord?, unparseableField: String?): ParsedRow {
        if (unparseableField != null) {
            return ParsedRow(rowNum, null, null, null, null, null, 0, false, "Invalid format or unparseable value for: $unparseableField", false)
        }

        if (record == null) {
            return ParsedRow(rowNum, null, null, null, null, null, 0, false, "Unknown parsing error", false)
        }

        // Column-specific skip values check
        val skipChecks = mapOf(
            config.getAnesthesiologistColumn() to record.anesthesiologistName,
            config.getSurgeonColumn() to record.surgeon,
            config.getProcedureColumn() to record.procedure,
            config.getSiteColumn() to record.site,
            config.getOrColumn() to record.orName
        )

        for ((col, value) in skipChecks) {
            val specificSkips = config.columnSkipValues[col]
            if (specificSkips != null && value != null && specificSkips.contains(value.trim().lowercase())) {
                return ParsedRow(rowNum, record, null, null, null, null, 0, false, "Ignored value for $col: $value", false)
            }
        }

        // Duplicate Case ID check
        if (seenCaseIds.containsKey(record.id)) {
            val original = seenCaseIds[record.id]!!
            if (original != record) {
                return ParsedRow(rowNum, record, null, null, null, null, 0, false, "Duplicate Case_ID with different data", false)
            }
            return ParsedRow(rowNum, record, null, null, null, null, 0, false, "Duplicate Case_ID", false)
        }
        seenCaseIds[record.id] = record
        
        // Procedure-specific padding check
        val procPadding = record.procedure?.uppercase()?.let { proc ->
            config.procedurePaddings[proc]
        }

        val startPad = procPadding?.first ?: config.startTimePad
        val anesthesiaStart = record.csvStart.minusMinutes(startPad.toLong())

        val endPad = procPadding?.second ?: config.endTimePad
        val anesthesiaEnd = record.csvEnd.plusMinutes(endPad.toLong())

        // CFO Evolution Phase 1: Use Physical Occupancy for Stacking/Packing
        val occupancyStart = record.patientIn.minusMinutes(startPad.toLong())
        val occupancyEnd = record.patientOut.plusMinutes(endPad.toLong())

        var valid = true
        var invalidReason: String? = null
        val duration = ChronoUnit.MINUTES.between(occupancyStart, occupancyEnd)

        if (duration < 0) { // occupancyEnd before occupancyStart (cross-midnight)
            valid = false
            invalidReason = "Cross-midnight duration (ignored per requirement)"
        } else if (duration == 0L) {
            valid = false
            invalidReason = "Zero duration"
        }

        if (valid && duration > 1440) { // Should not happen with LocalTime but for safety
            valid = false
            invalidReason = "Duration exceeds 24 hours"
        }

        if (!valid) {
            logger.info("Row {}: Invalid case. Reason: {}. Start: {}, End: {}", rowNum, invalidReason, occupancyStart, occupancyEnd)
        }

        // Check if outside absolute boundaries
        var outOfBounds = false
        val absStart = config.absoluteStartTime
        val absEnd = config.absoluteEndTime
        
        if (absStart != null && absEnd != null) {
            if (occupancyStart.isBefore(absStart) || occupancyEnd.isAfter(absEnd)) {
                outOfBounds = true
            }
        }

        return ParsedRow(rowNum, record, anesthesiaStart, anesthesiaEnd, occupancyStart, occupancyEnd, duration, valid, invalidReason, outOfBounds)
    }

    fun reset() {
        seenCaseIds.clear()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(CaseValidator::class.java)
    }
}
